### PR TITLE
Small refactor of the posteriors to support other models

### PR DESCRIPTION
**Pull request by johnamcleod**
**PR type:** enhancement / doc improvement

**Related issue(s)/PRs:** 

## Summary

**Proposed changes**

This PR refactors the posterior objects. The current posterior objects are tightly coupled to the SVGP model, by requiring both inducing points and variational parameters. This PR changes the constructor of the base class (`AbstractPosterior`) to accept the `alpha` and `Qinv` values (alongside the inducing points or training data points). This makes the base class sufficiently flexible to support the GPR, SGPR and VGP models, which can be added later.

The posterior object is not necessarily responsible for calculating `alpha` and `Qinv`. Subclasses of `AbstractPosterior` for different posteriors could have a `precompute` method like the SVGP `BasePosterior`, or the models could calculate the tensor values when creating a posterior object.

The base class for the SVGP posteriors is still `BasePosterior`. Updating the cached values has been left on the `BasePosterior`. 

This PR also includes a draft of a notebook about the caching functionality, with examples of the maths for the GPR, SGPR and (S)VGP models.

**What alternatives have you considered?**

### Minimal working example

See the timing example at the end of the new notebook.

### Release notes

Refactor the posterior base classes to support other models than the SVGP model.

**Fully backwards compatible:** yes

**If not, why is it worth breaking backwards compatibility:**

## PR checklist

- [ ] New features: code is well-documented
  - [ ] detailed docstrings (API documentation)
  - [x] notebook examples (usage demonstration)
- [x] The bug case / new feature is covered by unit tests
- [ ] Code has type annotations
- [x] Build checks
  - [x] I ran the black+isort formatter (`make format`)
  - [x] I locally tested that the tests pass (`make check-all`)
- [ ] Release management
  - [x] RELEASE.md updated with entry for this change
  - [ ] New contributors: I've added myself to CONTRIBUTORS.md

### CODE REVIEW

File: gpflow/posteriors.py [hunk 1]
```diff
@@ -123,13 +120,89 @@ def __init__(
...
+        if self.alpha is None or self.Qinv is None:
+            raise ValueError(
+                "Cache has not been precomputed yet. Call update_cache first or use fused_predict_f"
+            )
+        mean, cov = self._conditional_with_precompute(
+            Xnew, full_cov=full_cov, full_output_cov=full_output_cov
+        )
+        return self._add_mean_function(Xnew, mean), cov
+
+    @abstractmethod
```
Comment (st--) [inline gpflow/posteriors.py:174]: If it's abstract, why does it have an implementation?

File: gpflow/posteriors.py [hunk 2]
```diff
@@ -123,13 +120,89 @@ def __init__(
...
+        q_mu: tf.Tensor,
+        q_sqrt: tf.Tensor,
+        whiten: bool = True,
+        mean_function: Optional[mean_functions.MeanFunction] = None,
+        *,
+        precompute_cache: Optional[PrecomputeCacheType],
+    ):
+
+        super().__init__(kernel, inducing_variable, mean_function=mean_function)
         self.mean_function = mean_function
```
Comment (st--) [inline gpflow/posteriors.py:202]:
```suggestion
```

File: doc/source/notebooks/advanced/fast_predictions.pct.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
+# ---
+# jupyter:
```
Comment (st--) [inline doc/source/notebooks/advanced/fast_predictions.pct.py:2]: Still need to link this document from the notebook table of contents

**Review by st-- — COMMENTED** [review]
This generally seems fine; my main question is regarding follow-up PRs to enable support for GPR and SGPR: will we have to maintain duplicate code for all combinations of multi-output kernels and models? Is that something we could avoid, or inherent in the maths?

File: doc/source/notebooks/advanced/fast_predictions.pct.py (hunk 3 above)
Comment (johnamcleod) [reply doc/source/notebooks/advanced/fast_predictions.pct.py:2]: Good point - I have added this to the `intro.md` file.

File: gpflow/posteriors.py (hunk 1 above)
Comment (johnamcleod) [reply gpflow/posteriors.py]: An oversight. I have removed the decorator.

**Review by johnamcleod — COMMENTED** [review]
> will we have to maintain duplicate code for all combinations of multi-output kernels and models? Is that something we could avoid, or inherent in the maths?

There should be no need to have duplicate code to support the combinations of kernels and models.

**Review by st-- — APPROVED** [review]
lgtm

**Pull request merged**